server.js

```diff
@@ -57,8 +131,67 @@ app.post("/user/breached", function(req, res) {
   res.json({ info: "breach alert sent", emails: response });
 });
 
+// This object exists instead of inlining the env vars to make it easy
+// to abstract fetching API endpoints from the OAuth server (instead
+// of specifying them in the environment) in the future.
+const FxAOAuthUtils = {
+  get authorizationUri() { return process.env.OAUTH_AUTHORIZATION_URI },
+  get tokenUri() { return process.env.OAUTH_TOKEN_URI },
+  get profileUri() { return process.env.OAUTH_PROFILE_URI },
+};
+
+var FxAOAuth = new ClientOAuth2({
+  clientId: process.env.OAUTH_CLIENT_ID,
+  clientSecret: process.env.OAUTH_CLIENT_SECRET,
+  accessTokenUri: FxAOAuthUtils.tokenUri,
+  authorizationUri: FxAOAuthUtils.authorizationUri,
+  redirectUri: localServerURL + "/oauth/redirect",
+  scopes: ["profile:email"],
+});
+
+app.get("/oauth/init", function(req, res) {
+  // Set a random state string in a cookie so that we can verify
+  // the user when they're redirected back to us after auth.
+  let state = getStateString();
+  let uri = FxAOAuth.code.getUri({state});
+  req.session.state = state;
+  res.redirect(uri);
+});
+
+app.get('/oauth/redirect', function (req, res) {
+  if (!req.session.state) {
+    res.send("Who are you?");
+    return;
+  }
+  FxAOAuth.code.getToken(req.originalUrl, { state: req.session.state })
+    .then(function (user) {
+      popsicle.get({
```

Comment:
The `.catch()` is now in the better place, but I think this still has a weird side effect of any errors generated during the `popsicle.get(...).then(...)` won't throw up to the `.catch()` statement. Not sure if the right answer is to do a `return popsicle.get()` or how best to flatten the promises (or rethrow, or try and use `async`/`await`.

Reply:
FWIW, I'm investigating using async/await. I want to refactor this stuff in #16.